kep/9-capacity-scheduling-for-ml-dl-workloads-based-on-scheduler-framework/README.md

```diff
@@ -0,0 +1,208 @@
+- max: it will check the resource in the Prefilter of Pod Scheduling Cycle. The usage of max is aggregated based on the resource configurations of successfully scheduled pods.
+- min: it means the guaranteed resource of Quota.
+
+They act on different phases of pod lifecycle. So it maybe possible to merge them together.
+
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+We assume two elastic quotas are defined: QuotaA (min:4, max:6) and QuotaB (min:6, max:8), the quota unit is the number of GPUs. UserA in green color consumes QuotaA and UserB in red color consumes QuotaB. The entire cluster has 10 GPUs available hence the sum of quota min is equal to the cluster capacity.
+
+Initially, UserA consumes 2 GPUs and UserB consumes 3 GPUs. Both consumptions are within the reserved "min" quota range.
+![1](./1.png)
+
+
+Next, UserA consumes 4 GPUs reaching the "min" quota of QuotaA. UserB still consumes 3 GPUs. The cluster still has sufficient GPU resources hence no resource stealing happens.
+![2](./2.png)
+
+
+Assuming UserA attempts to use the rest of 3 unused GPUs in the cluster,  he can only use 2 more GPUs since his quota "max" is 6 GPUs. Note that these 2 GPUs are "borrowed" from UserB's "Guaranteed" quota since UserB does not use up it. 
+![3](./3.png)
+
+Later, if UserB demands 8 GPUs, 5 more compared to his current consumption, his request cannot be immediately satisfied since only one unused GPU left in the cluster. Since UserB's "min" quota is 6 GPUs, the preemption algorithm ensures that three GPUs will be returned from UserA.  However, no more GPUs can be assigned to UserB since UserA's "min" quota has to be honored as well. In the end, UserA consumes 4 GPUs and UserB consumes 6 GPUs, satisfying each user's "min" quota.
+![4](./4.png)
+
+## Design Details
+
+### Extention point
+
+#### PreFilter
+We will check if the (Pod.request + Quota.Allocated) is less than Quota.Max. If not, the scheduling cycle of PodA will fail. 
+
+#### PostFilter
+We will reimplement the method `FindCandidates` in `defaultPreempt`. The original `FindCandidates` method selects all the pods with the lower priority than the preemptor’s priority as potential victims in a node. 
+In this proposal, the victim selection logic needs to be changed in the following cases.
+- if Preemptor.Request + Quota.allocated > Quota.min: It means that its min or guaranteed resource is used or `borrowed` by other Quota. So we will ask the quotas who's allocated is more than max (it means thest quotas borrow the resource from other quota)
```

Comment:
"if _Preemptor.Request + Quota.allocated > Quota.min_: "

Should it be opposite?    " _Preemptor.Request + Quota.allocated <= Quota.min_"

"_So we will ask the quotas who's allocated is more than max (it means thest quotas borrow the resource from other quota)_"

"more than **min**"?

The sentence can be rephrased  as "a victim will be chosen from a namespace that allocates more resources than its min, i.e., borrowing resources from other namespaces" ?

How?  

1. Based on our offline discussion, it's randomly chosen.  
2. If evicting a single pod in namespace cannot release enough resources for the pod to be scheduled, will it continue to evict more pods in the same namespace or choose another one?

Reply:
+1 to Yuan. I think here the "if clause" in the 2 bullets should be reversed.

Reply:
> If evicting a single pod in namespace cannot release enough resources for the pod to be scheduled, will it continue to evict more pods in the same namespace or choose another one?
----
The current design does not differentiate between different namespaces. Just Inter-Quota or Intra-Quota. If `if Preemptor.Request + Quota.allocated > Quota.min`, the preempt will happen Inter-Quota. All the pods belongs to other Quotas that allocates more resources than its min in a node will be selected as the potential victims. The subsequent selection strategy remains the current logic: priority, PDB, createtime.

Reply:
> Should it be opposite? " Preemptor.Request + Quota.allocated <= Quota.min"
---
Done thanks